src/utils.ts

```diff
@@ -1,6 +1,196 @@
+  );
+}
+
+function encodeValue(
+  operator: string,
+  value: string,
+  key?: string
+): string {
+  const encodedValue =
+    operator === "+" || operator === "#"
+      ? encodeReserved(value)
+      : encodeUnreserved(value);
+
+  return key
+    ? `${encodeUnreserved(key)}=${encodedValue}`
+    : encodedValue;
+}
+
+function isDefined(value: unknown): boolean {
+  return value !== undefined && value !== null;
+}
+
+function isKeyOperator(operator: string): boolean {
+  return [";", "&", "?"].includes(operator);
+}
+
+function getValues(
+  context: Context,
+  operator: string,
+  key: string,
+  modifier?: string
+): string[] {
+  const value = context[key];
+  const result: string[] = [];
+
+  if (isDefined(value) && value !== "") {
+    if (
+      typeof value === "string" ||
+      typeof value === "number" ||
+      typeof value === "boolean"
+    ) {
+      let stringValue = value.toString();
+
+      if (modifier && modifier !== "*") {
+        stringValue = stringValue.substring(
+          0,
+          parseInt(modifier, 10)
+        );
+      }
+
+      result.push(
+        encodeValue(
+          operator,
+          stringValue,
+          isKeyOperator(operator) ? key : undefined
+        )
+      );
+    } else if (modifier === "*") {
+      if (Array.isArray(value)) {
+        value.filter(isDefined).forEach((item) => {
+          result.push(
+            encodeValue(
+              operator,
+              item.toString(),
+              isKeyOperator(operator) ? key : undefined
+            )
+          );
+        });
+      } else {
+        Object.entries(value as Record<string, unknown>).forEach(
+          ([subKey, subValue]) => {
+            if (isDefined(subValue)) {
+              result.push(
+                encodeValue(operator, String(subValue), subKey)
+              );
+            }
+          }
+        );
+      }
+    } else {
+      const tmp: string[] = [];
+
+      if (Array.isArray(value)) {
+        value.filter(isDefined).forEach((item) => {
+          tmp.push(encodeValue(operator, item.toString()));
+        });
+      } else {
+        Object.entries(value as Record<string, unknown>).forEach(
+          ([subKey, subValue]) => {
+            if (isDefined(subValue)) {
+              tmp.push(encodeUnreserved(subKey));
+              tmp.push(encodeValue(operator, String(subValue)));
+            }
+          }
+        );
+      }
+
+      if (isKeyOperator(operator)) {
+        result.push(`${encodeUnreserved(key)}=${tmp.join(",")}`);
+      } else if (tmp.length > 0) {
+        result.push(tmp.join(","));
+      }
+    }
+  } else {
+    if (operator === ";" && isDefined(value)) {
+      result.push(encodeUnreserved(key));
+    } else if (value === "" && ["&", "?"].includes(operator)) {
+      result.push(`${encodeUnreserved(key)}=`);
+    } else if (value === "") {
+      result.push("");
+    }
+  }
+
+  return result;
+}
+
+function parseTemplate(template: string): Template {
```

Comment:
Put this in a directory under utils with the license

https://github.com/bramstein/url-template/blob/main/LICENSE

> Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer